lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs

```diff
@@ -1323,6 +1326,48 @@ prop_balanceTransactionExistingTotalCollateral
   where
     pp = mockPParamsForBalancing
 
+-- If 'balanceTx' is able to balance a transaction, then repeating the attempt
+-- with all potential inputs removed (from the partial transaction and the
+-- UTxO set) should always fail with 'ErrBalanceTxUnableToCreateInput'.
+--
+-- Note that we /cannot/ expect 'balanceTx' to fail with
+-- 'ErrBalanceTxUnableToCreateInput' in all situations where there are no
+-- potential inputs available, since:
+--
+-- 1. only at most one failure condition is ever reported by 'balanceTx';
+-- 2. there can be multiple competing failure conditions;
+-- 3. the order in which failure conditions are checked is unspecified.
+--
+prop_balanceTransactionUnableToCreateInput
+    -- TODO: Test with all recent eras [ADP-2997]
+    :: forall era. era ~ Write.BabbageEra
+    => Wallet
+    -> PartialTx era
+    -> StdGenSeed
+    -> Property
+prop_balanceTransactionUnableToCreateInput w ptx seed =
+    case r0 of
+        Right _balancedTx -> r1 === Left ErrBalanceTxUnableToCreateInput
+        Left _someFailure -> property True
+    & cover 1
+        (isRight r0)
+        "isRight r0"
+    & checkCoverage
```

Comment:
If the `Left` case is uninteresting and we take care to ensure some `Right` coverage, then wouldn't using `==>` be both easier and clearer? 🤔 

E.g.:
```haskell
isRight r0 ==> r1 === Left ErrBalanceTxUnableToCreateInput
```

Reply:
> If the `Left` case is uninteresting and we take care to ensure some `Right` coverage, then wouldn't using `==>` be both easier and clearer? 🤔
> 
> E.g.:
> 
> ```haskell
> isRight r0 ==> r1 === Left ErrBalanceTxUnableToCreateInput
> ```

I avoided using the `==>` combinator, as on average only a very small number of cases result in a successfully balanced transaction (around 2%). This leads to test flakiness, as QuickCheck often gives up if the LHS of `==>` is repeatedly unsatisfied.

I've pushed some further commits to this PR that introduce a `BalanceTxArgs` record type with a `Success` combinator. The generator for `Success BalanceTxArgs` guarantees to always generate arguments that lead to balanced transactions.